Add support for DKIM headers in e-mails

ywarnier opened the pull request.

jmontoyaa commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5254203&bid=17216401)

jmontoyaa commented on app/config/mail.conf.dist.php:
```diff
@@ -19,3 +19,12 @@
+// to match the following selector
+// Also, using SMTP_UNIQUE_SENDER is required if users have e-mails from different domains
+$platform_email['DKIM'] = 0; //enable DKIM by setting this to 1
+$platform_email['DKIM_SELECTOR'] = 'chamilo'; // an indicator of the application sending the e-mail through this specific DKIM key
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Line exceeds maximum limit of 120 characters; contains 130 characters](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5254203)

jmontoyaa commented on app/config/mail.conf.dist.php:
```diff
@@ -19,3 +19,12 @@
+$platform_email['DKIM_SELECTOR'] = 'chamilo'; // an indicator of the application sending the e-mail through this specific DKIM key
+$platform_email['DKIM_DOMAIN'] = 'mydomain.com'; //the domain for e-mail sending, not necessarily api_get_path(WEB_PATH)
+$platform_email['DKIM_PRIVATE_KEY_STRING'] = ''; //the private key in a string format
+$platform_email['DKIM_PRIVATE_KEY'] = ''; //the private key as the path to a file. The file needs to be accessible to PHP!
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Line exceeds maximum limit of 120 characters; contains 122 characters](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5254203)

jmontoyaa commented:
This changes deletes **all the folder** main/inc/lib/phpmailer! The phpmailer library was highly changed in Chamilo to handle specific situations.

ywarnier commented:
@jmontoyaa 
Can you point me to the specific situations handled? I think there was something about unique sender, but this, I thought, would be handled in api.lib.php (api_mail_html()).

ywarnier commented:
Here is a (simplified and commented) diff between v5.1 (the one we used to have in Chamilo with no dependency management) and v5.1 from Git from the PHPMailer project.

As we will see, I believe everything is now dealt with by the new version of PHPMailer we depend on.

```
diff --git a/class.phpmailer.php b/tmp/class.phpmailer.php
index 602f121..4a1b8e7
--- a/class.phpmailer.php
+++ b/tmp/class.phpmailer.php
@@ -451,7 +451,7 @@ class PHPMailer {
    */
   private function AddAnAddress($kind, $address, $name = '') {
     if (!preg_match('/^(to|cc|bcc|ReplyTo)$/', $kind)) {
-      echo 'Invalid recipient array: ' . kind;
+      error_log('Invalid recipient array: ' . $kind);
       return false;
     }
     $address = trim($address);
@@ -461,7 +461,7 @@ class PHPMailer {
       if ($this->exceptions) {
         throw new phpmailerException($this->Lang('invalid_address').': '.$address);
       }
-      echo $this->Lang('invalid_address').': '.$address;
+      error_log($this->Lang('invalid_address').': '.$address);
       return false;
     }
     if ($kind != 'ReplyTo') {
@@ -493,7 +493,7 @@ class PHPMailer {
       if ($this->exceptions) {
         throw new phpmailerException($this->Lang('invalid_address').': '.$address);
       }
-      echo $this->Lang('invalid_address').': '.$address;
+      error_log($this->Lang('invalid_address').': '.$address);
       return false;
     }
     $this->From = $address;
@@ -583,7 +583,6 @@ class PHPMailer {
       if ($this->exceptions) {
         throw $e;
       }
-      echo $e->getMessage()."\n";
       return false;
     }
   }
```
Until here, we just changed (and I believe this was made by the PHPMailer team in a minor version rather than by us) echo into error_log. In version 6, the method $this->setError() is used, which handles this just fine.
```
@@ -595,11 +594,34 @@ class PHPMailer {
    * @access protected
    * @return bool
    */
-  protected function SendmailSend($header, $body) {
-    if ($this->Sender != '') {
-      $sendmail = sprintf("%s -oi -f %s -t", escapeshellcmd($this->Sendmail), escapeshellarg($this->Sender));
+  protected function SendmailSend($header, $body)
+  {
+      if (!(is_file($this->Sendmail) and is_executable($this->Sendmail))) {
+          throw new phpmailerException(
+              $this->lang('execute').$this->Sendmail,
+              self::STOP_CRITICAL
+          );
+      }
+      if (!empty($this->Sender) and $this->validateAddress($this->Sender)) {
+          if ($this->Mailer == 'qmail') {
+              $sendmail = sprintf(
+                  '%s -f%s',
+                  escapeshellcmd($this->Sendmail),
+                  escapeshellarg($this->Sender)
+              );
+          } else {
+              $sendmail = sprintf(
+                  '%s -oi -f%s -t',
+                  escapeshellcmd($this->Sendmail),
+                  escapeshellarg($this->Sender)
+              );
+          }
+      } else {
+          if ($this->Mailer == 'qmail') {
+              $sendmail = sprintf('%s', escapeshellcmd($this->Sendmail));
           } else {
-      $sendmail = sprintf("%s -oi -t", escapeshellcmd($this->Sendmail));
+              $sendmail = sprintf('%s -oi -t', escapeshellcmd($this->Sendmail));
+          }
       }
     if ($this->SingleTo === true) {
       foreach ($this->SingleToArray as $key => $val) {
```
I doubt those changes were made by us.
```
@@ -649,7 +671,7 @@ class PHPMailer {
     $to = implode(', ', $toArr);
 
     $params = sprintf("-oi -f %s", $this->Sender);
-    if ($this->Sender != '' && strlen(ini_get('safe_mode'))< 1) {
+    if (!empty($this->Sender) and !ini_get('safe_mode') and $this->validateAddress($this->Sender)) {
       $old_from = ini_get('sendmail_from');
       ini_set('sendmail_from', $this->Sender);
       if ($this->SingleTo === true && count($toArr) > 1) {
@@ -705,7 +727,12 @@ class PHPMailer {
     if(!$this->SmtpConnect()) {
       throw new phpmailerException($this->Lang('smtp_connect_failed'), self::STOP_CRITICAL);
     }
-    $smtp_from = ($this->Sender == '') ? $this->From : $this->Sender;
+
+    if (!empty($this->Sender) and $this->validateAddress($this->Sender)) {
+        $smtp_from = $this->Sender;
+    } else {
+        $smtp_from = $this->From;
+    }
     if(!$this->smtp->Mail($smtp_from)) {
       throw new phpmailerException($this->Lang('from_failed') . $smtp_from, self::STOP_CRITICAL);
     }
```
Same here. This seems to be much more internal work than an integration change for Chamilo
```
@@ -1367,7 +1394,7 @@ class PHPMailer {
       if ($this->exceptions) {
         throw $e;
       }
-      echo $e->getMessage()."\n";
+      error_log($e->getMessage()."\n");
       if ( $e->getCode() == self::STOP_CRITICAL ) {
         return false;
       }
@@ -1461,18 +1488,23 @@ class PHPMailer {
       if (!is_readable($path)) {
         throw new phpmailerException($this->Lang('file_open') . $path, self::STOP_CONTINUE);
       }
-      if (function_exists('get_magic_quotes')) {
-        function get_magic_quotes() {
-          return false;
-        }
-      }
-      if (PHP_VERSION < 6) {
         $magic_quotes = get_magic_quotes_runtime();
+        if ($magic_quotes) {
+            if (version_compare(PHP_VERSION, '5.3.0', '<')) {
                 set_magic_quotes_runtime(0);
+            } else {
+                ini_set('magic_quotes_runtime', 0);
+            }
         }
         $file_buffer = file_get_contents($path);
         $file_buffer = $this->EncodeString($file_buffer, $encoding);
-      if (PHP_VERSION < 6) { set_magic_quotes_runtime($magic_quotes); }
+        if ($magic_quotes) {
+            if (version_compare(PHP_VERSION, '5.3.0', '<')) {
+                set_magic_quotes_runtime($magic_quotes);
+            } else {
+                ini_set('magic_quotes_runtime', $magic_quotes);
+            }
+        }
         return $file_buffer;
     } catch (Exception $e) {
       $this->SetError($e->getMessage());
```
This looks internal, but it used to be something special for PHP < 5.3, whereas now the dependencies manager should take care of this and Chamilo only supports PHP 5.6 or superior (and it is pretty likely it will not support PHP 5.6 fully starting with the next version, which is the one that includes the new PHPMailer.
```
@@ -1552,7 +1584,10 @@ class PHPMailer {
     // Try to select the encoding which should produce the shortest output
     if (strlen($str)/3 < $x) {
       $encoding = 'B';
-      if (function_exists('mb_strlen') && $this->HasMultiBytes($str)) {
+      // Modified by Ivan Tcholakov, 24-JAN-2010.
+      //if (function_exists('mb_strlen') && $this->HasMultiBytes($str)) {
+      if ($this->HasMultiBytes($str)) {
+      //
         // Use a custom function which correctly encodes and wraps long
         // multibyte strings without breaking lines within a character
         $encoded = $this->Base64EncodeWrapMB($str);
```
Finally, this seems to be a customization by us (Ivan contributed to Dokeos in the late 2000's and early 2010).
However, it just gets rid of a check on the installation of the MultiByte PHP extension, which is now a suggestion in phpmailer's own composer.json, so I can only suppose it deals well with the fact that the extension might not always be installed. Either way, the comments in the code change by Ivan do not allow us to understand exactly why it was made.
```
@@ -1580,12 +1615,14 @@ class PHPMailer {
    * @param string $str multi-byte text to wrap encode
    * @return bool
    */
-  public function HasMultiBytes($str) {
+    public function HasMultiBytes($str)
+    {
         if (function_exists('mb_strlen')) {
             return (strlen($str) > mb_strlen($str, $this->CharSet));
         } else { // Assume no multibytes (we can't handle without mbstring functions anyway)
             return false;
         }
+
     }
 
   /**
```
Just a coding style change.
```
@@ -1601,7 +1638,10 @@ class PHPMailer {
     $end = "?=";
     $encoded = "";
 
-    $mb_length = mb_strlen($str, $this->CharSet);
+    // Modified by Ivan Tcholakov, 24-JAN-2010.
+    //$mb_length = mb_strlen($str, $this->CharSet);
+    $mb_length = api_strlen($str, $this->CharSet);
+    //
     // Each line must have length <= 75, including $start and $end
     $length = 75 - strlen($start) - strlen($end);
     // Average multi-byte ratio
@@ -1614,7 +1654,10 @@ class PHPMailer {
 
       do {
         $offset = $avgLength - $lookBack;
-        $chunk = mb_substr($str, $i, $offset, $this->CharSet);
+        // Modified by Ivan Tcholakov, 24-JAN-2010.
+        //$chunk = mb_substr($str, $i, $offset, $this->CharSet);
+        $chunk = api_substr($str, $i, $offset, $this->CharSet);
+        //
         $chunk = base64_encode($chunk);
         $lookBack++;
       }
```
Again slight changes regarding the MB extension. If my memory serves, the internal api_substr() and api_strlen() functions were created to allow the interchangeability between MultiByte and intl extensions, but I'm not sure about it. If it is, it is definitely something I would like to rely first on for a version of PHPMailer that is 10 years more updated. Then if needed, we'll update.
```
@@ -1724,15 +1767,15 @@ class PHPMailer {
 
     switch (strtolower($position)) {
       case 'phrase':
-        $encoded = preg_replace("/([^A-Za-z0-9!*+\/ -])/e", "'='.sprintf('%02X', ord('\\1'))", $encoded);
+        $encoded = @preg_replace("/([^A-Za-z0-9!*+\/ -])/e", "'='.sprintf('%02X', ord('\\1'))", $encoded);
         break;
       case 'comment':
-        $encoded = preg_replace("/([\(\)\"])/e", "'='.sprintf('%02X', ord('\\1'))", $encoded);
+        $encoded = @preg_replace("/([\(\)\"])/e", "'='.sprintf('%02X', ord('\\1'))", $encoded);
       case 'text':
       default:
         // Replace every high ascii, control =, ? and _ characters
         //TODO using /e (equivalent to eval()) is probably not a good idea
-        $encoded = preg_replace('/([\000-\011\013\014\016-\037\075\077\137\177-\377])/e',
+        $encoded = @preg_replace('/([\000-\011\013\014\016-\037\075\077\137\177-\377])/e',
               "'='.sprintf('%02X', ord('\\1'))", $encoded);
         break;
     }
```
This is just ignoring some errors that were making things messy. Hopefully, this is not necessary anymore.
```

So all in all, I am convinced that the customizations we added to PHPMailer were not that essential anymore and I am confident we can depend on the new version directly as downloaded by composer.
